---
# Predictive Business Loan Default Analysis Model
---

**Author**: Dhevina Tailor

**Contact**: [dhevinatailor@gmail.com](mailto:dhevinatailor@gmail.com)

**Date**: October 20, 2023

## Table of Contents
---
1. [Introduction: Data Cleaning](#introduction)
2. [Key Questions](#keyquestions)
3. [Feature Overview](#featureoverview)
4. [Loading and Setup](#loadingandsetup)
5. [Assessment](#assessment)
6. [Cleaning](#cleaning)
7. [Feature Handling Strategies to Mitigate Data Leakage](#features)
8. [Saving the Data](#saving)

## Introduction: Data Cleaning <a class="anchor" id="introduction"></a>
---

Before embarking on the analysis phase, it's imperative to ensure that our dataset is in optimal condition. Data cleaning is a critical step in this process. Conducting a first pass over my data to identify data quality issues, feature engineering opportunities and any other notable observations regarding data preprocessing.

**Refining the Raw Data:**
Our initial dataset may contain inconsistencies, missing values, or outliers. To prepare it for analysis, we'll meticulously clean the data. This includes addressing missing values, removing duplicates, and standardizing formats.

**Ensuring Data Integrity:**
A clean dataset is the foundation for reliable and accurate analysis. We'll take steps to ensure data integrity and consistency, allowing us to draw meaningful conclusions in the subsequent analysis phase.


## Key Questions <a class="anchor" id="keyquestions"></a>
---

#### The analysis aims to answer these fundamental questions:

**Using machine learning, how might we accurately predict the future status of a business loan payment (paid or default), to improve risk management and reduce financial losses?**

This question serves as the foundation for the analysis, guiding us in the exploration of the loan dataset derived from the Small Business Administration (SBA).

**Specifically tailored to our loan default dataset, we will address the following key factors:**

1. **Data Quality:** Inconsistencies, missing values, or outliers in the dataset that could affect the accuracy of our loan default prediction model.
2. **Data Integrity:** Ensure the data's integrity and consistency throughout the analysis process, particularly in the context of loan default analysis?
3. **Data Preparation:** Steps required to refine and prepare the dataset for robust loan default prediction and in-depth analysis?

These key questions collectively shape our data cleaning and analysis efforts, ensuring that the loan default dataset is ready for the development of an effective loan default prediction model.

## Feature Overview <a class="anchor" id="featureoverview"></a>
---

<style>
  table {
    width: 100%;
    border-collapse: collapse;
  }

  th, td {
    border: 1px solid black;
    padding: 8px;
    text-align: left;
  }

  th {
    background-color: #f2f2f2;
    font-weight: bold;
  }
</style>

| **#** | **Variable Name**    | **Description**                                        |
|-------|----------------------|--------------------------------------------------------|
| **1** | LoanNr_ChkDgt        | Identifier Primary key                                 |
| **2** | Name                 | Borrower name                                         |
| **3** | City                 | Borrower city                                         |
| **4** | State                | Borrower state                                        |
| **5** | Zip                  | Borrower zip code                                     |
| **6** | Bank                 | Bank name                                             |
| **7** | BankState            | Bank state                                            |
| **8** | NAICS                | North American industry classification system code   |
| **9** | ApprovalDate         | Date SBA commitment issued                            |
| **10** | ApprovalFY           | Fiscal year of commitment                             |
| **11** | Term                 | Loan term in months                                   |
| **12** | NoEmp                | Number of business employees                          |
| **13** | NewExist             | 1 = Existing business, 2 = New business               |
| **14** | CreateJob            | Number of jobs created                                |
| **15** | RetainedJob          | Number of jobs retained                               |
| **16** | FranchiseCode        | Franchise code, (00000 or 00001) = No franchise       |
| **17** | UrbanRural           | 1 = Urban, 2 = rural, 0 = undefined                   |
| **18** | RevLineCr            | Revolving line of credit: Y = Yes, N = No, 0 = undefined|
| **19** | LowDoc               | LowDoc Loan Program: Y = Yes, N = No                 |
| **20** | ChgOffDate           | The date when a loan is declared to be in default    |
| **21** | DisbursementDate     | Disbursement date                                     |
| **22** | DisbursementGross    | Amount disbursed                                      |
| **23** | BalanceGross         | Gross amount outstanding                              |
| **24** | MIS_Status           | Loan status charged off = CHGOFF, Paid in full =PIF |
| **25** | ChgOffPrinGr         | Charged-off amount                                    |
| **26** | GrAppv               | Gross amount of loan approved by bank                |
| **27** | SBA_Appv             | SBA’s guaranteed amount of approved loan              |



Description of the first two digits of NAICS.

<style>
  table {
    width: 100%;
    border-collapse: collapse;
  }

  th {
    border: 1px solid black;
    padding: 8px;
    text-align: left;
    background-color: #f2f2f2;
    font-weight: bold;
  }
</style>

| **Sector** | **Description**                                           |
|------------|-----------------------------------------------------------|
| 11         | Agriculture, forestry, fishing, and hunting                |
| 21         | Mining, quarrying, and oil and gas extraction             |
| 22         | Utilities                                                  |
| 23         | Construction                                               |
| 31–33      | Manufacturing                                              |
| 42         | Wholesale trade                                            |
| 44–45      | Retail trade                                               |
| 48–49      | Transportation and warehousing                            |
| 51         | Information                                                |
| 52         | Finance and insurance                                      |
| 53         | Real estate and rental and leasing                         |
| 54         | Professional, scientific, and technical services           |
| 55         | Management of companies and enterprises                    |
| 56         | Administrative and support and waste management and remediation services |
| 61         | Educational services                                       |
| 62         | Health care and social assistance                          |
| 71         | Arts, entertainment, and recreation                        |
| 72         | Accommodation and food services                            |
| 81         | Other services (except public administration)              |
| 92         | Public administration                                      |


## Loading and Setup <a class="anchor" id="loadingandsetup"></a>

In [1]:
# importing required library
import pandas as pd

### Get Data

In [2]:
# Load the dataset
df = pd.read_csv("SBAnational.csv")

/var/folders/2j/c8kl0vnx1cd5zbnyjj5331fw0000gn/T/ipykernel_93264/24265492.py:2: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("SBAnational.csv")


In [3]:
# Create a dictionary to store column names and descriptions
column_descriptions = {
    'LoanNr_ChkDgt': 'Identifier Primary key',
    'Name': 'Borrower name',
    'City': 'Borrower city',
    'State': 'Borrower state',
    'Zip': 'Borrower zip code',
    'Bank': 'Bank name',
    'BankState': 'Bank state',
    'NAICS': 'North American industry classification system code',
    'ApprovalDate': 'Date SBA commitment issued',
    'ApprovalFY': 'Fiscal year of commitment',
    'Term': 'Loan term in months',
    'NoEmp': 'Number of business employees',
    'NewExist': '1 = Existing business, 2 = New business',
    'CreateJob': 'Number of jobs created',
    'RetainedJob': 'Number of jobs retained',
    'FranchiseCode': 'Franchise code, (00000 or 00001) = No franchise',
    'UrbanRural': '1 = Urban, 2 = rural, 0 = undefined',
    'RevLineCr': 'Revolving line of credit: Y = Yes, N = No',
    'LowDoc': 'LowDoc Loan Program: Y = Yes, N = No',
    'ChgOffDate': 'The date when a loan is declared to be in default',
    'DisbursementDate': 'Disbursement date',
    'DisbursementGross': 'Amount disbursed',
    'BalanceGross': 'Gross amount outstanding',
    'MIS_Status': 'Loan status charged off = CHGOFF, Paid in full = PIF',
    'ChgOffPrinGr': 'Charged-off amount',
    'GrAppv': 'Gross amount of loan approved by the bank',
    'SBA_Appv': "SBA’s guaranteed amount of the approved loan"
}

# Function to retrieve the description of a column
def get_column_description(column_name):
    return column_descriptions.get(column_name, 'Description not available for this column.')

## Assessment <a class="anchor" id="assessment"></a>

In [4]:
# Set display options to show all columns and rows
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [5]:
# Print the shape of the dataset
df.shape

(899164, 27)

In [6]:
print(f'There are {df.shape[0]} rows and {df.shape[1]} columns in this dataset')

There are 899164 rows and 27 columns in this dataset


Next, we inspect the first rows of data to get a feel for the structure.

In [7]:
# Print the first 5 rows of the dataset
df.head()

LoanNr_ChkDgt                           Name          City State    Zip  \
0     1000014003                 ABC HOBBYCRAFT    EVANSVILLE    IN  47711   
1     1000024006    LANDMARK BAR & GRILLE (THE)     NEW PARIS    IN  46526   
2     1000034009          WHITLOCK DDS, TODD M.   BLOOMINGTON    IN  47401   
3     1000044001  BIG BUCKS PAWN & JEWELRY, LLC  BROKEN ARROW    OK  74012   
4     1000054004    ANASTASIA CONFECTIONS, INC.       ORLANDO    FL  32801   

                            Bank BankState   NAICS ApprovalDate ApprovalFY  \
0               FIFTH THIRD BANK        OH  451120    28-Feb-97       1997   
1                1ST SOURCE BANK        IN  722410    28-Feb-97       1997   
2        GRANT COUNTY STATE BANK        IN  621210    28-Feb-97       1997   
3  1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97       1997   
4        FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97       1997   

   Term  NoEmp  NewExist  CreateJob  RetainedJob  FranchiseCode  UrbanRural  \
0    84      4       2.0          0            0              1           0   
1    60      2       2.0          0            0              1           0   
2   180      7       1.0          0            0              1           0   
3    60      2       1.0          0            0              1           0   
4   240     14       1.0          7            7              1           0   

  RevLineCr LowDoc ChgOffDate DisbursementDate DisbursementGross BalanceGross  \
0         N      Y        NaN        28-Feb-99       $60,000.00        $0.00    
1         N      Y        NaN        31-May-97       $40,000.00        $0.00    
2         N      N        NaN        31-Dec-97      $287,000.00        $0.00    
3         N      Y        NaN        30-Jun-97       $35,000.00        $0.00    
4         N      N        NaN        14-May-97      $229,000.00        $0.00    

  MIS_Status ChgOffPrinGr        GrAppv      SBA_Appv  
0      P I F       $0.00    $60,000.00    $48,000.00   
1      P I F       $0.00    $40,000.00    $32,000.00   
2      P I F       $0.00   $287,000.00   $215,250.00   
3      P I F       $0.00    $35,000.00    $28,000.00   
4      P I F       $0.00   $229,000.00   $229,000.00

In [8]:
# Display the info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 899164 entries, 0 to 899163
Data columns (total 27 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   LoanNr_ChkDgt      899164 non-null  int64  
 1   Name               899150 non-null  object 
 2   City               899134 non-null  object 
 3   State              899150 non-null  object 
 4   Zip                899164 non-null  int64  
 5   Bank               897605 non-null  object 
 6   BankState          897598 non-null  object 
 7   NAICS              899164 non-null  int64  
 8   ApprovalDate       899164 non-null  object 
 9   ApprovalFY         899164 non-null  object 
 10  Term               899164 non-null  int64  
 11  NoEmp              899164 non-null  int64  
 12  NewExist           899028 non-null  float64
 13  CreateJob          899164 non-null  int64  
 14  RetainedJob        899164 non-null  int64  
 15  FranchiseCode      899164 non-null  int64  
 16  Ur

In [9]:
# Display the data types of the dataset
df.dtypes

LoanNr_ChkDgt          int64
Name                  object
City                  object
State                 object
Zip                    int64
Bank                  object
BankState             object
NAICS                  int64
ApprovalDate          object
ApprovalFY            object
Term                   int64
NoEmp                  int64
NewExist             float64
CreateJob              int64
RetainedJob            int64
FranchiseCode          int64
UrbanRural             int64
RevLineCr             object
LowDoc                object
ChgOffDate            object
DisbursementDate      object
DisbursementGross     object
BalanceGross          object
MIS_Status            object
ChgOffPrinGr          object
GrAppv                object
SBA_Appv              object
dtype: object

In [10]:
# Check the number of unique values in each column
df.nunique()

LoanNr_ChkDgt        899164
Name                 779583
City                  32581
State                    51
Zip                   33611
Bank                   5802
BankState                56
NAICS                  1312
ApprovalDate           9859
ApprovalFY               70
Term                    412
NoEmp                   599
NewExist                  3
CreateJob               246
RetainedJob             358
FranchiseCode          2768
UrbanRural                3
RevLineCr                18
LowDoc                    8
ChgOffDate             6448
DisbursementDate       8472
DisbursementGross    118859
BalanceGross             15
MIS_Status                2
ChgOffPrinGr          83165
GrAppv                22128
SBA_Appv              38326
dtype: int64

In [11]:
# Get a statistical summary of the dataset
df.describe()

LoanNr_ChkDgt            Zip          NAICS           Term  \
count   8.991640e+05  899164.000000  899164.000000  899164.000000   
mean    4.772612e+09   53804.391241  398660.950146     110.773078   
std     2.538175e+09   31184.159152  263318.312759      78.857305   
min     1.000014e+09       0.000000       0.000000       0.000000   
25%     2.589758e+09   27587.000000  235210.000000      60.000000   
50%     4.361439e+09   55410.000000  445310.000000      84.000000   
75%     6.904627e+09   83704.000000  561730.000000     120.000000   
max     9.996003e+09   99999.000000  928120.000000     569.000000   

               NoEmp       NewExist      CreateJob    RetainedJob  \
count  899164.000000  899028.000000  899164.000000  899164.000000   
mean       11.411353       1.280404       8.430376      10.797257   
std        74.108196       0.451750     236.688165     237.120600   
min         0.000000       0.000000       0.000000       0.000000   
25%         2.000000       1.000000       0.000000       0.000000   
50%         4.000000       1.000000       0.000000       1.000000   
75%        10.000000       2.000000       1.000000       4.000000   
max      9999.000000       2.000000    8800.000000    9500.000000   

       FranchiseCode     UrbanRural  
count  899164.000000  899164.000000  
mean     2753.725933       0.757748  
std     12758.019136       0.646436  
min         0.000000       0.000000  
25%         1.000000       0.000000  
50%         1.000000       1.000000  
75%         1.000000       1.000000  
max     99999.000000       2.000000

### Applicability of `df.describe()`

The `df.describe()` method, which provides summary statistics, can be applied to the following columns:

- **Term**: The 'Term' column likely represents the loan term in months, which is a numerical feature. You can use `df.describe()` to obtain statistics like mean, standard deviation, and quartiles for this column. It helps us understand the distribution of loan terms and their central tendencies.

- **NoEmp**: This column contains numerical data, representing the number of business employees. `df.describe()` is well-suited for generating statistics like the mean, standard deviation, and quartiles for this column. It helps us understand the central tendency and spread of the employee count in the dataset.

- **CreateJob** and **RetainedJob**: Both of these columns are likely related to job creation and retention. They are numerical features, and `df.describe()` can be used to derive statistics on these metrics.

For the following columns, `df.describe()` is not applicable:

- **LoanNr_ChkDgt**: This column likely serves as an identifier or primary key for each loan entry. It contains unique values and is not suitable for generating meaningful numerical statistics.

- **Zip**: Zip codes are categorical data representing geographical locations and postal codes. `df.describe()` is not applicable to zip codes as it's designed for numerical statistics, not categorical data.

- **NAICS**: NAICS codes represent industry categories and classifications. These are categorical variables, and `df.describe()` is not appropriate for generating numerical statistics for this type of data.

- **NewExist**: This column contains categorical values (1 and 2) indicating whether a business is existing (1) or new (2). It doesn't provide meaningful numerical statistics, and using `df.describe()` on categorical data like this is not suitable.

- **FranchiseCode**: Franchise codes are categorical values representing franchise status. `df.describe()` is not suitable for generating numerical statistics for this type of data.

- **UrbanRural**: This column contains categorical values (0, 1, and 2) indicating urban, rural, or undefined locations. `df.describe()` is not applicable to this type of categorical data.

Understanding the applicability of `df.describe()` for specific columns is essential for utilizing this method effectively in data analysis, ensuring that it provides meaningful insights based on the nature of the data.

In [12]:
# Check the number of null values in each column
df.isnull().sum()

LoanNr_ChkDgt             0
Name                     14
City                     30
State                    14
Zip                       0
Bank                   1559
BankState              1566
NAICS                     0
ApprovalDate              0
ApprovalFY                0
Term                      0
NoEmp                     0
NewExist                136
CreateJob                 0
RetainedJob               0
FranchiseCode             0
UrbanRural                0
RevLineCr              4528
LowDoc                 2582
ChgOffDate           736465
DisbursementDate       2368
DisbursementGross         0
BalanceGross              0
MIS_Status             1997
ChgOffPrinGr              0
GrAppv                    0
SBA_Appv                  0
dtype: int64

In [13]:
# Check for duplicate rows
df.duplicated().sum()

0

## Cleaning <a class="anchor" id="cleaning"></a>

We will clean the following columns: 'DisbursementGross', 'BalanceGross', 'ChgOffPrinGr', 'GrAppv', and 'SBA_Appv'.

- 'DisbursementGross' represents the amount of money disbursed.
- 'BalanceGross' represents the balance amount.
- 'ChgOffPrinGr' is the charged-off principal amount.
- 'GrAppv' is the approved gross amount.
- 'SBA_Appv' is the SBA-approved amount.

We will remove any dollar signs, commas, and spaces from these columns to convert the values into numerical format.

In [14]:
columns_to_clean = ['DisbursementGross', 'BalanceGross', 'ChgOffPrinGr', 'GrAppv', 'SBA_Appv']

for column in columns_to_clean:
    df[column] = df[column].replace('[$, ]', '', regex=True).astype(float)

In [15]:
# Display the data types of the dataset
df.dtypes

LoanNr_ChkDgt          int64
Name                  object
City                  object
State                 object
Zip                    int64
Bank                  object
BankState             object
NAICS                  int64
ApprovalDate          object
ApprovalFY            object
Term                   int64
NoEmp                  int64
NewExist             float64
CreateJob              int64
RetainedJob            int64
FranchiseCode          int64
UrbanRural             int64
RevLineCr             object
LowDoc                object
ChgOffDate            object
DisbursementDate      object
DisbursementGross    float64
BalanceGross         float64
MIS_Status            object
ChgOffPrinGr         float64
GrAppv               float64
SBA_Appv             float64
dtype: object

**Data Type Conversion: From Object to Float**

- The columns we cleaned ('DisbursementGross', 'BalanceGross', 'ChgOffPrinGr', 'GrAppv', and 'SBA_Appv') were of data type 'object.' This data type is a generic container for strings and mixed types of data. However, when working with numerical values, using 'float' is preferred.

- The 'float' data type represents real numbers and is suitable for storing decimal and floating-point values. It allows us to perform mathematical operations, aggregations, and statistical analyses on the data.

By converting these columns to 'float,' we have made the data more accessible for a wide range of data analysis and modeling tasks. We can now proceed to perform further analysis on these columns and gain valuable insights from the cleaned and properly formatted numerical data.

In [16]:
# Check the percentage of missing values in each column
df.isnull().sum() / len(df) * 100

LoanNr_ChkDgt         0.000000
Name                  0.001557
City                  0.003336
State                 0.001557
Zip                   0.000000
Bank                  0.173383
BankState             0.174162
NAICS                 0.000000
ApprovalDate          0.000000
ApprovalFY            0.000000
Term                  0.000000
NoEmp                 0.000000
NewExist              0.015125
CreateJob             0.000000
RetainedJob           0.000000
FranchiseCode         0.000000
UrbanRural            0.000000
RevLineCr             0.503579
LowDoc                0.287156
ChgOffDate           81.905526
DisbursementDate      0.263356
DisbursementGross     0.000000
BalanceGross          0.000000
MIS_Status            0.222095
ChgOffPrinGr          0.000000
GrAppv                0.000000
SBA_Appv              0.000000
dtype: float64

In [17]:
# Create a copy of the original dataframe
df_clean = df.copy()

# Get the columns with more than 20% missing values
cols_to_drop = df_clean.columns[df_clean.isnull().sum() / len(df_clean) * 100 > 20]

# Drop the columns with more than 20% missing values
df_clean.drop(cols_to_drop, axis=1, inplace=True)

In [18]:
# Check the shape of the dataframe
df_clean.shape

(899164, 26)

### Data Cleaning Methodology

**Step 1: Calculate Missing Data Percentage**

We begin by calculating the percentage of missing values in each column. This involves using the `isnull()` method to create a Boolean DataFrame, summing these values for each column, and dividing by the total row count.

**Step 2: Create Clean DataFrame Copy**

We create a duplicate of the original DataFrame, `df`, named `df_clean`, to preserve the original data.

**Step 3: Identify Columns with >20% Missing Data**

We pinpoint and filter out columns in `df_clean` with over 20% missing data using Boolean indexing.

**Step 4: Drop High Missing Data Columns**

The code removes the identified columns with excessive missing data from `df_clean` using the `drop()` method.

**Step 5: Check DataFrame Dimensions**

We verify the shape of the resulting `df_clean` DataFrame to confirm the row and column count.

### Findings

1. The original DataFrame, `df`, contained 899,164 rows and 27 columns.

2. After the data cleaning process, `df_clean` retains the same row count but is reduced to 26 columns.

   - This means one column had over 20% missing data in the original DataFrame, and it was removed, resulting in a more manageable and complete dataset for analysis or modeling.


In [19]:
# Check the percentage of missing values in each column in descending order
(df_clean.isnull().sum() / len(df_clean) * 100).sort_values(ascending=False)

RevLineCr            0.503579
LowDoc               0.287156
DisbursementDate     0.263356
MIS_Status           0.222095
BankState            0.174162
Bank                 0.173383
NewExist             0.015125
City                 0.003336
State                0.001557
Name                 0.001557
LoanNr_ChkDgt        0.000000
UrbanRural           0.000000
GrAppv               0.000000
ChgOffPrinGr         0.000000
BalanceGross         0.000000
DisbursementGross    0.000000
CreateJob            0.000000
FranchiseCode        0.000000
RetainedJob          0.000000
NoEmp                0.000000
Term                 0.000000
ApprovalFY           0.000000
ApprovalDate         0.000000
NAICS                0.000000
Zip                  0.000000
SBA_Appv             0.000000
dtype: float64

We will now direct our attention to a column that showcases a higher rate of missing values compared to the rest of the dataset:

- **RevLineCr:** Approximately 50% of the rows are missing data in this column.
- **LowDoc:** Approximately 29% of the rows are missing data in this column.
- **DisbursementDate:** Approximately 26% of the rows are missing data in this column.
- **MIS_Status:** Approximately 22% of the rows are missing data in this column.
- **BankState:** Approximately 17% of the rows are missing data in this column.
- **Bank:** Approximately 17% of the rows are missing data in this column.
- **City:** Approximately 0.3% of the rows are missing data in this column.
- **State:** Approximately 0.2% of the rows are missing data in this column.
- **Name:** Approximately 0.2% of the rows are missing data in this column.

In [20]:
column_name = 'RevLineCr'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: RevLineCr
Column Description: Revolving line of credit: Y = Yes, N = No


### Interpretation of the "RevLineCr" Column:

The "RevLineCr" column in the dataset represents the presence or absence of a revolving line of credit. A revolving line of credit is a type of credit arrangement where a borrower is granted a specific credit limit and can borrow, repay, and borrow again up to that limit. It is often associated with credit cards and similar financial products.

**Interpretation:**

- **Y (Yes)**: Indicates that the borrower has a revolving line of credit.
- **N (No)**: Indicates that the borrower does not have a revolving line of credit.

**Significance:**

The presence of a revolving line of credit can be an important factor in assessing a borrower's financial stability and creditworthiness. It may affect the borrower's ability to manage multiple lines of credit and their financial responsibilities.

**Timing of Variable in the Loan Application Process:**

The information in the "RevLineCr" column is available at the time of the loan application, prior to the actual loan disbursement.

**Decision on Data Handling:**

As the "RevLineCr" column provides insights into the borrower's financial situation at the time of loan application, it is relevant for loan prediction and analysis. Therefore, it is recommended to **retain** this column for further analysis and modeling.


### Identifying Unique Values in the 'RevLineCr' Column

Let's begin by examining all the unique values in the 'RevLineCr' column to understand its content. This is crucial as we intend to retain only 'Y' and 'N' values, and any other values will be dropped.

In [21]:
# Get unique values in the 'RevLineCr' column
unique_values = df_clean['RevLineCr'].unique()

# Print the unique values
print(unique_values)

['N' '0' 'Y' 'T' nan '`' ',' '1' 'C' '3' '2' 'R' '7' 'A' '5' '.' '4' '-'
 'Q']


In [22]:
# Filter rows in the 'RevLineCr' column that contain 'Y' or 'N'
df_clean = df_clean[df_clean['RevLineCr'].isin(['Y', 'N', '0'])]

# Reset the index
df_clean = df_clean.reset_index(drop=True)

In [23]:
# Get unique values after filtering
unique_values_after = df_clean['RevLineCr'].unique()

# Print the unique values after filtering
print("\nUnique Values After Filtering:")
print(unique_values_after)


Unique Values After Filtering:
['N' '0' 'Y']


In [24]:
column_name = 'LowDoc'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: LowDoc
Column Description: LowDoc Loan Program: Y = Yes, N = No


### Interpretation of the "LowDoc" Column:

The "LowDoc" column in the dataset is an indicator of whether the loan applied for falls under the LowDoc Loan Program. The LowDoc Loan Program is a specific lending program designed for small businesses to simplify the loan application process, especially for loans of relatively small amounts. However, it's important to note that as of September 2021, the U.S. Small Business Administration (SBA) eliminated the LowDoc Loan Program.

**Interpretation:**

- **Y (Yes)**: Indicates that the loan is classified under the LowDoc Loan Program before its elimination.
- **N (No)**: Indicates that the loan does not fall under the LowDoc Loan Program.

**Significance:**

Prior to its elimination, the LowDoc Loan Program aimed to reduce the documentation and paperwork requirements for small business loans, making it more accessible to small business owners. However, it's important to consider that the program is no longer in effect.

**Timing of Variable in the Loan Application Process:**

The information in the "LowDoc" column is available at the time of the loan application, prior to the actual loan disbursement. Loans classified under the LowDoc program were subject to its streamlined documentation requirements until the program's elimination in September 2021.

**Decision on Data Handling:**

As of September 2021, the "LowDoc" column may contain data from loans classified under the LowDoc Loan Program before its elimination. Given that the LowDoc Loan Program has been eliminated, the "LowDoc" column may no longer provide relevant information for predicting loan defaults. Therefore, it is recommended to **exclude** this column for further analysis and modeling.

In [25]:
# drop the LowDoc column
df_clean.drop('LowDoc', axis=1, inplace=True)

In [26]:
df_clean.head()

LoanNr_ChkDgt                           Name          City State    Zip  \
0     1000014003                 ABC HOBBYCRAFT    EVANSVILLE    IN  47711   
1     1000024006    LANDMARK BAR & GRILLE (THE)     NEW PARIS    IN  46526   
2     1000034009          WHITLOCK DDS, TODD M.   BLOOMINGTON    IN  47401   
3     1000044001  BIG BUCKS PAWN & JEWELRY, LLC  BROKEN ARROW    OK  74012   
4     1000054004    ANASTASIA CONFECTIONS, INC.       ORLANDO    FL  32801   

                            Bank BankState   NAICS ApprovalDate ApprovalFY  \
0               FIFTH THIRD BANK        OH  451120    28-Feb-97       1997   
1                1ST SOURCE BANK        IN  722410    28-Feb-97       1997   
2        GRANT COUNTY STATE BANK        IN  621210    28-Feb-97       1997   
3  1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97       1997   
4        FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97       1997   

   Term  NoEmp  NewExist  CreateJob  RetainedJob  FranchiseCode  UrbanRural  \
0    84      4       2.0          0            0              1           0   
1    60      2       2.0          0            0              1           0   
2   180      7       1.0          0            0              1           0   
3    60      2       1.0          0            0              1           0   
4   240     14       1.0          7            7              1           0   

  RevLineCr DisbursementDate  DisbursementGross  BalanceGross MIS_Status  \
0         N        28-Feb-99            60000.0           0.0      P I F   
1         N        31-May-97            40000.0           0.0      P I F   
2         N        31-Dec-97           287000.0           0.0      P I F   
3         N        30-Jun-97            35000.0           0.0      P I F   
4         N        14-May-97           229000.0           0.0      P I F   

   ChgOffPrinGr    GrAppv  SBA_Appv  
0           0.0   60000.0   48000.0  
1           0.0   40000.0   32000.0  
2           0.0  287000.0  215250.0  
3           0.0   35000.0   28000.0  
4           0.0  229000.0  229000.0

In [27]:
column_name = 'DisbursementDate'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: DisbursementDate
Column Description: Disbursement date


### Interpretation of the "DisbursementDate" Column:

The "DisbursementDate" column in the dataset provides information about the date on which a loan was disbursed or paid out to the borrower. This date marks the point at which the borrower gains access to the funds and can start using them for their intended business purposes.

**Interpretation:**

- The values in this column represent specific dates when loans were disbursed to borrowers.

**Significance:**

While "DisbursementDate" carries significance in understanding the timing of loan disbursement, it may not be directly relevant to predicting loan default. In a real-case scenario for loan default prediction, this information is typically not available.

**Timing of Variable in the Loan Application Process:**

The "DisbursementDate" variable is relevant during the loan disbursement phase, which occurs after the loan application approval process. It marks the point when the borrower receives the approved loan amount.

**Decision on Data Handling:**

Considering that "DisbursementDate" does not contribute to predicting loan default and may not be present in real-case scenarios when making predictions, it is reccomended that we **exclude** this column from the dataset. This streamlines the analysis, focusing on factors more directly associated with loan default prediction.


In [28]:
# drop the DisbursementDate column
df_clean.drop('DisbursementDate', axis=1, inplace=True)

In [29]:
df_clean.head()

LoanNr_ChkDgt                           Name          City State    Zip  \
0     1000014003                 ABC HOBBYCRAFT    EVANSVILLE    IN  47711   
1     1000024006    LANDMARK BAR & GRILLE (THE)     NEW PARIS    IN  46526   
2     1000034009          WHITLOCK DDS, TODD M.   BLOOMINGTON    IN  47401   
3     1000044001  BIG BUCKS PAWN & JEWELRY, LLC  BROKEN ARROW    OK  74012   
4     1000054004    ANASTASIA CONFECTIONS, INC.       ORLANDO    FL  32801   

                            Bank BankState   NAICS ApprovalDate ApprovalFY  \
0               FIFTH THIRD BANK        OH  451120    28-Feb-97       1997   
1                1ST SOURCE BANK        IN  722410    28-Feb-97       1997   
2        GRANT COUNTY STATE BANK        IN  621210    28-Feb-97       1997   
3  1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97       1997   
4        FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97       1997   

   Term  NoEmp  NewExist  CreateJob  RetainedJob  FranchiseCode  UrbanRural  \
0    84      4       2.0          0            0              1           0   
1    60      2       2.0          0            0              1           0   
2   180      7       1.0          0            0              1           0   
3    60      2       1.0          0            0              1           0   
4   240     14       1.0          7            7              1           0   

  RevLineCr  DisbursementGross  BalanceGross MIS_Status  ChgOffPrinGr  \
0         N            60000.0           0.0      P I F           0.0   
1         N            40000.0           0.0      P I F           0.0   
2         N           287000.0           0.0      P I F           0.0   
3         N            35000.0           0.0      P I F           0.0   
4         N           229000.0           0.0      P I F           0.0   

     GrAppv  SBA_Appv  
0   60000.0   48000.0  
1   40000.0   32000.0  
2  287000.0  215250.0  
3   35000.0   28000.0  
4  229000.0  229000.0

In [30]:
column_name = 'MIS_Status'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: MIS_Status
Column Description: Loan status charged off = CHGOFF, Paid in full = PIF


### Interpretation of the "MIS_Status" Column (Target Variable):

The "MIS_Status" column in the dataset serves as the primary target variable. It indicates the status of a loan, classifying it as either "Charged Off" (CHGOFF) or "Paid in Full" (PIF).

**Interpretation:**

- **Charged Off (CHGOFF)**: This category signifies that the loan has been declared as "charged off," typically indicating a significant delinquency or default.
- **Paid in Full (PIF)**: This category denotes that the loan has been successfully paid off in full by the borrower.

**Significance:**

The "MIS_Status" column is of utmost importance in understanding the loan's ultimate outcome. It allows us to categorize loans into two critical groups: those that resulted in default ("Charged Off") and those that were successfully repaid ("Paid in Full"). This information is vital for predictive modeling, enabling the assessment of factors that contribute to loan default.

**Timing of Variable in the Loan Application Process:**

The "MIS_Status" variable becomes known after the loan's term has progressed, and it reflects the outcome of the borrower's ability to meet their repayment obligations. It is a post-disbursement variable.

**Decision on Data Handling:**

As the primary target variable for predicting loan default, the "MIS_Status" column is valuable. Therefore, it is recommended to **retain** this column and utilize it as the focal point for building the predictive model to assess the loan default risk.


In [31]:
# We replace 'P|F' with 'Paid' and 'CHGOFF' with 'Default' to ensure consistency in the dataset.
df_clean['MIS_Status'] = df_clean['MIS_Status'].replace(['P I F'], 'Paid')
df_clean['MIS_Status'] = df_clean['MIS_Status'].replace(['CHGOFF'], 'Default')

### Identifying Unique Values in the 'MIS_Status' Column

Let's begin by examining all the unique values in the 'MIS_Status' column to understand its content. This is crucial as we intend to retain only 'Paid' and 'Default' values, and any other values will be dropped.

In [32]:
# Get unique values in the 'RevLineCr' column
unique_values = df_clean['MIS_Status'].unique()

# Print the unique values
print(unique_values)

['Paid' 'Default' nan]


In [33]:
# Drop rows with NaN values in the 'MIS_Status' column
df_clean = df_clean.dropna(subset=['MIS_Status'])

In [34]:
# Get unique values in the 'RevLineCr' column
unique_values = df_clean['MIS_Status'].unique()

# Print the unique values
print(unique_values)

['Paid' 'Default']


This results in two values, 'Paid' and 'Default'.

In [35]:
column_name = 'Name'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: Name
Column Description: Borrower name


In [37]:
column_name = 'Zip'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: Zip
Column Description: Borrower zip code


### Interpretation of the "ZIP" Column

In our dataset, the "Zip" column contains postal Zip codes associated with our borrowers or businesses, offering insights into their geographic locations. However, it's worth noting that we already have separate columns for "State" and "City," providing more detailed location information. As a result, the "Zip" column may seem redundant for our analysis.

**Significance:**

While Zip codes can be valuable for identifying geographic regions, in our specific case, their significance might be diminished due to the availability of more specific location data in the "State" and "City" columns.

**Timing of Variable in the Loan Application Process:**

The "Zip" variable, like other location-related data, remains relevant throughout the loan application process for assessing geographic factors.

**Decision on Data Handling:**

Given that we possess more detailed geographic information through the "State" and "City" columns, it seems reasonable to consider **excluding** the "Zip" column from our analysis. This would help us avoid redundancy and streamline our modeling process. However, the ultimate decision should be based on whether ZIP code data adds unique value to our specific analysis or if it can be safely omitted.


In [38]:
# drop the Zip column
df_clean.drop('Zip', axis=1, inplace=True)

In [39]:
column_name = 'BankState'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: BankState
Column Description: Bank state


### Interpretation of the "BankState" Column:

The "BankState" column in the dataset indicates the state in which the bank associated with the loan's approval is located. It provides information about the geographical location of the lending institution responsible for approving the loan.

**Interpretation:**

- The values in this column represent the two-letter abbreviations of U.S. states where the banks are headquartered.

**Significance:**

The "BankState" column can be significant for various analytical purposes. It may offer insights into regional banking trends, potential regulatory variations across states, and lender distribution.

**Timing of Variable in the Loan Application Process:**

The "BankState" variable is relevant throughout the loan application process, from the initial application to approval. It reflects the geographical origin of the lending institution.

**Decision on Data Handling:**

The inclusion of the "BankState" column is not of utmost importance but it is recommended to **retain** this column for exploratory purposes.

In [40]:
column_name = 'Bank'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: Bank
Column Description: Bank name


### Interpretation of the "Bank" Column:

The "Bank" column in the dataset represents the name of the bank associated with the loan's approval. It provides information about the lending institution responsible for approving and disbursing the loan.

**Interpretation:**

- The values in this column includes the names of banks and financial institutions involved in the loan approval process.

**Significance:**

The "Bank" column is significant for understanding the specific lending institutions involved in the loan application process. It may offer insights into the diversity of lenders, their individual policies, and their role in providing financial support to borrowers.

**Timing of Variable in the Loan Application Process:**

The "Bank" variable is relevant throughout the loan application process, from the initial application to approval and disbursement. It reflects the specific lending institution chosen by the borrower or the lender responsible for the loan.

**Decision on Data Handling:**

The inclusion of the "Bank" column is not of utmost importance but it is recommended to **retain** this column for exploratory purposes.

In [41]:
column_name = 'City'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: City
Column Description: Borrower city


### Interpretation of the "City" Column:

The "City" column in the dataset represents the city associated with the borrower's location. It provides information about the geographical location of the businesses or individuals applying for loans.

**Interpretation:**

- The values in this column includes the names of cities where borrowers are located.

**Significance:**

The "City" column is significant for understanding the geographical distribution of loan applicants. It offers insights into the regional concentration of borrowers and may help identify location-specific patterns, economic conditions, and local market trends.

**Timing of Variable in the Loan Application Process:**

The "City" variable is relevant throughout the loan application process, from the initial application to approval and disbursement. It reflects the geographical origin of the borrower and the location of the business.

**Decision on Data Handling:**

The inclusion of the "City" column can be valuable. It is recommended to **retain** this column for exploratory purposes.

In [42]:
column_name = 'State'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: State
Column Description: Borrower state


### Interpretation of the "State" Column:

The "State" column in the dataset represents the U.S. state associated with the borrower's location. It provides information about the geographical location of small businesses applying for loans.

**Interpretation:**

- The values in this column includes the two-letter abbreviations of U.S. states to denote the location of the borrower.

**Significance:**

The "State" column is significant for understanding the geographical distribution of loan applicants and allows for regional categorization. It offers insights into the diversity of borrowers across different states and may help identify location-specific economic conditions, regulatory factors, and market trends.

**Timing of Variable in the Loan Application Process:**

The "State" variable is relevant throughout the loan application process, from the initial application to approval and disbursement. It reflects the geographical origin of the borrower and the location of the business.

**Decision on Data Handling:**

The inclusion of the "State" column in your dataset can be valuable. It is recommended to **retain** this column for exploratory purposes.

In [43]:
column_name = 'Name'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: Name
Column Description: Borrower name


### Interpretation of the "Name" Column:

The "Name" column in the dataset represents the name or identifier of the borrower or business applying for the loan. It provides information about the entity/individual seeking financial support.

**Interpretation:**

- The values in this column includes the names of borrowers (businesses).

**Significance:**

The "Name" column serves to identify and categorize businesses in the dataset. While it is a fundamental component of the loan application information, its specific relevance to loan outcomes or default prediction may be limited.

**Timing of Variable in the Loan Application Process:**

The "Name" variable is relevant throughout the loan application process, from the initial application to approval and disbursement. It reflects the identity of the borrower or business entity.

**Decision on Data Handling:**

For the specific purpose of loan default prediction, the "Name" column may not significantly contribute. It is recommended to **retain** this column for exploratory purposes.

In [44]:
df_clean.shape

(877343, 23)

## Exploring Columns for Data Cleaning

We'll explore all the columns in the dataset. This step is instrumental in data cleaning, as it enables us to identify and address any data inconsistencies or anomalies.

In [45]:
column_name = 'NAICS'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: NAICS
Column Description: North American industry classification system code


### Interpretation of the "NAICS" Column

The "NAICS" (North American Industry Classification System) column in the dataset contains codes that classify the type of industry or business that the borrower or entity represents. These codes provide information about the industry or sector to which the borrower belongs.

**Interpretation:**

- The values in this column are NAICS codes representing different industries or sectors.

**Significance:**

The "NAICS" column is crucial for categorizing and understanding the nature of the businesses in the dataset. It offers insights into the diversity of industries seeking financial support through loans. The specific industry of a borrower can be an important factor in loan outcomes and default prediction.

**Timing of Variable in the Loan Application Process:**

The "NAICS" variable is relevant throughout the loan application process, as it reflects the industry or sector of the borrower or business. This information is considered during the initial application, approval, and disbursement stages of the loan.

**Decision on Data Handling:**

For the purpose of loan default prediction, the "NAICS" column can be a valuable feature. The industry or sector to which a business belongs can impact its financial stability and likelihood of default. It is recommended to **retain** this column for predictive modeling and analysis, as it can provide meaningful insights into loan outcomes.


In [46]:
column_name = 'NewExist'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: NewExist
Column Description: 1 = Existing business, 2 = New business


### Interpretation of the "NewExist" Column

The "NewExist" column in the dataset is used to categorize businesses based on their status as either new or existing entities. This classification helps in understanding the lifecycle stage of the businesses seeking loans.

**Interpretation:**

- The "NewExist" column contains values that categorize businesses as either new (1) or existing (2) entities.

**Significance:**

The "NewExist" variable is significant in classifying businesses based on their stage of development. It provides insights into whether the borrowers are startups or established businesses. This information can be relevant for assessing loan risk and default prediction.

**Timing of Variable in the Loan Application Process:**

The "NewExist" variable is pertinent during the entire loan application process. It reflects the stage of development of the business or entity, which can influence loan decisions, including approval and terms.

**Decision on Data Handling:**

For the purpose of loan default prediction, the "NewExist" column is a valuable feature. The lifecycle stage of a business can be a crucial factor in assessing its financial stability and likelihood of default. Therefore, it is recommended to **retain** this column for predictive modeling and analysis, as it can provide meaningful insights into loan outcomes.


### Identifying Unique Values in the 'NewExist' Column

Let's begin by examining all the unique values in the 'NewExist' column to understand its content. This is crucial as we intend to retain only '1' and '2' values, and any other values will be dropped.

In [47]:
# Get unique values in the 'NewExist' column
unique_values_newexist = df_clean['NewExist'].unique()

# Print the unique values
print(unique_values_newexist)

[ 2.  1.  0. nan]


In [48]:
# Keep rows with values 1 and 2 in the 'NewExist' column
df_clean = df_clean[df_clean['NewExist'].isin([1, 2])]

In [49]:
# Get unique values in the 'NewExist' column
unique_values_newexist = df_clean['NewExist'].unique()

# Print the unique values
print(unique_values_newexist)

[2. 1.]


In [50]:
df_clean.shape

(876192, 23)

In [51]:
column_name = 'FranchiseCode'
description = get_column_description(column_name)
print(f'Column Name: {column_name}')
print(f'Column Description: {description}')

Column Name: FranchiseCode
Column Description: Franchise code, (00000 or 00001) = No franchise


### Interpretation of the "FranchiseCode" Column

The "FranchiseCode" column in the dataset contains numerical codes that indicate whether a business is a franchise or not. These codes provide information about whether the borrower's business operates as a franchise.

**Interpretation:**

- The "FranchiseCode" column contains numerical codes where specific values represent franchises, while others indicate non-franchise businesses.

**Significance:**

The "FranchiseCode" variable is crucial for distinguishing between businesses that are part of a franchise and those that are independently owned. This information is significant for assessing the structure and business model of the borrowers, which can have implications for loan outcomes and default prediction.

**Timing of Variable in the Loan Application Process:**

The "FranchiseCode" variable is relevant during the entire loan application process. It reflects whether the business is affiliated with a franchise or operates independently, which can influence loan decisions.

**Decision on Data Handling:**

For the purpose of loan default prediction, the "FranchiseCode" column can be a valuable feature. Understanding whether a business is a franchise or independent can provide insights into loan risk and borrower characteristics. It is recommended to **retain** this column for predictive modeling and analysis, as it can offer valuable information for assessing loan outcomes.


### Identifying Unique Values in the 'FranchiseCode' Column

In this section, we will explore the unique values present in the 'FranchiseCode' column. This exploration is essential as we have retained values starting with '11', '21', '22', and other specified two-digit prefixes, along with '0' and '1', while any other values have been excluded from our analysis.

In [52]:
# Get unique values in the 'FranchiseCode' column
unique_values_franchisecode = df_clean['FranchiseCode'].unique()

# Print the unique values
print(unique_values_franchisecode)

[    1     0 15100 ...  2899 18701 15930]


In [53]:
# Define the list of valid two-digit prefixes
valid_prefixes = ['11', '21', '22', '23', '31', '32', '33', '42', '44', '45', '48', '49', '51', '52', '53', '54', '55', '56', '61', '62', '71', '72', '81', '92']

# Include one-digit values '0' and '1'
valid_prefixes += ['0', '1']

# Convert 'FranchiseCode' to string type
df_clean['FranchiseCode'] = df_clean['FranchiseCode'].astype(str)

# Filter rows based on the 'FranchiseCode' column
df_clean = df_clean[df_clean['FranchiseCode'].str[:2].isin(valid_prefixes) | (df_clean['FranchiseCode'].isin(['00000', '00001']))]


In [54]:
# Get unique values in the 'NewExist' column
unique_values_franchisecode = df_clean['FranchiseCode'].unique()

# Print the unique values
print(unique_values_franchisecode)

['1' '0' '22470' '21425' '21400' '23997' '51570' '81612' '81965' '23379'
 '22475' '21420' '23000' '72590' '71456' '44081' '52000' '21780' '42520'
 '45455' '81630' '53250' '56480' '62343' '52956' '55120' '61400' '49952'
 '4516' '56837' '45479' '33570' '61901' '33550' '62753' '61900' '21790'
 '48450' '31150' '9202' '22678' '81602' '55512' '56885' '52018' '48000'
 '32074' '55185' '54600' '53165' '45150' '22680' '71658' '21800' '53657'
 '33750' '22676' '51720' '55960' '45890' '71685' '11245' '2105' '56920'
 '54100' '45505' '53650' '33170' '52707' '81550' '45100' '72515' '62907'
 '32077' '81800' '42533' '31243' '45495' '31300' '32195' '5650' '52875'
 '32130' '55524' '61401' '55193' '61897' '21660' '48260' '33470' '42599'
 '11250' '21870' '44350' '51778' '31415' '52300' '56880' '52855' '4250'
 '45955' '44725' '51560' '2357' '56925' '45915' '49855' '9201' '48400'
 '62346' '62751' '71158' '81597' '54400' '61390' '62100' '42580' '55955'
 '81570' '53680' '72683' '23375' '62760' '61603' '56575' '

**Data Cleaning Note:**

The data in the "FranchiseCode" column has been cleaned to include only rows with 'FranchiseCode' values that start with specific two-digit prefixes or are '00000' or '00001'. This cleaning step ensures that the data in this column is consistent and conforms to the specified criteria.

The unique values in the "FranchiseCode" column have been reviewed to confirm that the data has been appropriately cleaned and filtered.


In [55]:
# Drop missing values in object columns
df_clean.dropna(inplace=True)

# Check the percentage of missing values in each column
missing_percentage = df_clean.isnull().sum() / len(df_clean) * 100

# Display the missing percentage for each column
print("Percentage of Missing Values in Each Column:")
print(missing_percentage)

Percentage of Missing Values in Each Column:
LoanNr_ChkDgt        0.0
Name                 0.0
City                 0.0
State                0.0
Bank                 0.0
BankState            0.0
NAICS                0.0
ApprovalDate         0.0
ApprovalFY           0.0
Term                 0.0
NoEmp                0.0
NewExist             0.0
CreateJob            0.0
RetainedJob          0.0
FranchiseCode        0.0
UrbanRural           0.0
RevLineCr            0.0
DisbursementGross    0.0
BalanceGross         0.0
MIS_Status           0.0
ChgOffPrinGr         0.0
GrAppv               0.0
SBA_Appv             0.0
dtype: float64


We have successfully addressed all missing values in the dataset, rendering it clean and prepared for Exploratory Data Analysis (EDA).

In [56]:
# Let's check the dimensions of the cleaned dataframe after handling missing values

df_clean.shape

(836136, 23)

## Feature Handling Strategies to Mitigate Data Leakage <a class="anchor" id="features"></a>

#### Post-Loan Features: Managing Data After Loan Approval

After a thorough examination of the remaining dataset columns, we have identified the following post-loan features that should be omitted. These features pertain to events occurring after the loans have been disbursed:

1. LoanNr_ChkDgt
2. ApprovalDate
3. ApprovalFY
4. Term
5. CreateJob
6. RetainedJob
7. ChgOffDate (Previously dropped)
8. DisbursementDate (Previously dropped)
9. DisbursementGross
10. BalanceGross
11. ChgOffPrinGr
12. GrAppv

In [57]:
# List of columns to drop
columns_to_drop = [
    'LoanNr_ChkDgt', 'ApprovalDate', 'ApprovalFY', 'Term', 'CreateJob',
    'RetainedJob', 'DisbursementGross',
    'BalanceGross', 'ChgOffPrinGr', 'GrAppv'
]

# Drop the columns from the DataFrame
df_clean.drop(columns=columns_to_drop, inplace=True)

We removed these columns from the dataset for the following reasons:

1. **LoanNr_ChkDgt**: This column contains loan numbers or identifiers, which don't really help us predict loan defaults. They're more for record-keeping.

2. **ApprovalDate** and **ApprovalFY**: These columns provide information about when a loan was approved. However, this timing isn't directly relevant to predicting loan defaults, as loan approval decisions are typically made before the actual loan is granted.

3. **Term**: The term of the loan, which indicates the repayment period, isn't a significant factor in predicting loan defaults. It's more related to the loan's structure.

4. **CreateJob** and **RetainedJob**: These columns show the number of jobs created or retained due to a loan. While this is important information for other purposes, it's not directly linked to predicting loan defaults.

7. **ChgOffDate** and **DisbursementDate** (previously dropped): These dates are associated with specific loan events, like charge-offs and disbursements, which occur after a loan has already defaulted. Including them in predictions would reveal information about the outcome, leading to data leakage.

8. **DisbursementGross**: This column represents the loan disbursement amount, and its relevance to predicting loan defaults may be limited or highly correlated with other variables.

9. **BalanceGross**: The remaining loan balance doesn't help predict defaults, and including it could lead to data leakage if it reflects the loan's performance.

10. **ChgOffPrinGr**: This is the charged-off principal amount resulting from loan defaults. Including it would introduce data leakage, as it's a consequence of the event we're trying to predict.

11. **GrAppv**: The gross amount approved might not be a strong predictor for loan defaults, or it could be highly correlated with other approved amounts.

Overall, these columns were dropped to avoid data leakage and concentrate on the most relevant features for predicting loan defaults.

In [58]:
# Verify the DataFrame's dimensions following the removal of post-loan features
df_clean.shape

(836136, 13)

In [59]:
print(f'To ensure data integrity and prevent data leakage, we removed the post-loan features, resulting in a remaining dataset with {df_clean.shape[0]} rows and {df_clean.shape[1]} columns.')

To ensure data integrity and prevent data leakage, we removed the post-loan features, resulting in a remaining dataset with 836136 rows and 13 columns.


In [60]:
# Display the names of the remaining columns in the cleaned dataset
df_clean.columns.tolist()

['Name',
 'City',
 'State',
 'Bank',
 'BankState',
 'NAICS',
 'NoEmp',
 'NewExist',
 'FranchiseCode',
 'UrbanRural',
 'RevLineCr',
 'MIS_Status',
 'SBA_Appv']

In [61]:
# Show a sample of the data
df_clean.head()

Name          City State  \
0                 ABC HOBBYCRAFT    EVANSVILLE    IN   
1    LANDMARK BAR & GRILLE (THE)     NEW PARIS    IN   
2          WHITLOCK DDS, TODD M.   BLOOMINGTON    IN   
3  BIG BUCKS PAWN & JEWELRY, LLC  BROKEN ARROW    OK   
4    ANASTASIA CONFECTIONS, INC.       ORLANDO    FL   

                            Bank BankState   NAICS  NoEmp  NewExist  \
0               FIFTH THIRD BANK        OH  451120      4       2.0   
1                1ST SOURCE BANK        IN  722410      2       2.0   
2        GRANT COUNTY STATE BANK        IN  621210      7       1.0   
3  1ST NATL BK & TR CO OF BROKEN        OK       0      2       1.0   
4        FLORIDA BUS. DEVEL CORP        FL       0     14       1.0   

  FranchiseCode  UrbanRural RevLineCr MIS_Status  SBA_Appv  
0             1           0         N       Paid   48000.0  
1             1           0         N       Paid   32000.0  
2             1           0         N       Paid  215250.0  
3             1           0         N       Paid   28000.0  
4             1           0         N       Paid  229000.0

### Duplicated Data

In [62]:
df_clean.duplicated().sum()

896

We are dropping duplicates in the dataset because multiple entries of the same information exist. These duplicates provide no additional value and can skew analysis results, making it essential to maintain a clean and accurate dataset for meaningful insights.

In [63]:
# Drop duplicate rows from the DataFrame
df_clean = df_clean.drop_duplicates()

In [64]:
# Check for duplicates

df_clean.duplicated().sum()

0

Data cleaning is complete. Let's now save this cleaned dataset for future use.

## Saving the data <a class="anchor" id="saving"></a>

Now that we have spent the time cleaning this dataset, we are going to save it for future use. We will be using our cleaned dataset to carry out some exploratory work to learn more about the dataset!

In [65]:
# Save the cleaned data to a csv file for future use

df_clean.to_csv('Cleaned_Loan_Data.csv', index=False)